vim-patch:vim-9.1.{0874-0876}: detect more filetypes

clason opened the pull request:
- **vim-patch:9.1.0874: filetype: karel files are not detected**
- **vim-patch:9.1.0875: filetype: hyprlang detection can be improved**
- **vim-patch:9.1.0876: filetype: openCL files are not recognized**

clason committed:
vim-patch:9.1.0874: filetype: karel files are not detected

Problem:  filetype: karel files are not detected
Solution: detect '*.kl' files as karel filetype,
          include syntax and filetype plugin
          (Kirill Morozov)

closes: vim/vim#16075

https://github.com/vim/vim/commit/fdac54d7bbf6d68a8bf741e734b86d0f1998ac86

Co-authored-by: Kirill Morozov <kirill@robotix.pro>
Co-authored-by: KnoP-01 <knosowski@graeffrobotics.de>

clason committed:
vim-patch:9.1.0875: filetype: hyprlang detection can be improved

Problem:  filetype: hyprlang detection can be improved
Solution: detect '/hypr/*.conf' files as hyprlang filetype,
          include basic syntax highlighting (Luca Saccarola)

fixes: vim/vim#15875
closes: vim/vim#16064

https://github.com/vim/vim/commit/a13bd294ab2d9ab38634c9ec51fa76205af6eb62

Co-authored-by: Luca Saccarola <github.e41mv@aleeas.com>

clason committed:
vim-patch:9.1.0876: filetype: openCL files are not recognized

Problem:  filetype: openCL files are not recognized
Solution: detect '*.cl' files as opencl or lisp filetype,
          include a opencl syntax and filetype plugin (Wu, Zhenyu)

closes: vim/vim#15825

https://github.com/vim/vim/commit/e2c27ca8eff7cc8ec852b531d5a7f328a343a761

Co-authored-by: Wu, Zhenyu <wuzhenyu@ustc.edu>

lewis6991 commented on runtime/lua/vim/filetype.lua:
```diff
@@ -2139,6 +2141,7 @@ local pattern = {
     ['%.git/worktrees/.*/config%.worktree$'] = 'gitconfig',
     ['/%.config/git/config$'] = 'gitconfig',
     ['/gitolite%-admin/conf/'] = starsetf('gitolite'),
+    ['/hypr/.*%.conf$'] = 'hyprlang',
```
Just a note that these patterns with `.*` are really difficult to optimise around.

If the `.*` was replaced with something like `[^/]*` then we could do theoretically break up these patterns more easily and do quicker matches on the filename first. This would make generating a trie much easier.

I did a prototype of generating a trie, but I had to exclude too many patterns like this and the perf increase was only around 15%. If we can somehow simplify these patterns (remove `.*`), then the perf increase would be much more.

Is there any scope for this or must we copy these patterns from vim verbatim?

clason replied:
Well, any deviation makes porting patterns harder, to the point where I will just stop. (Not blackmail; just projection.)

But from the point of view of the project, what matters is only the effective detection, so any optimization is fair game that does not affect that. (Unfortunately, the Vim tests tend to ignore the more complicated cases, so we can't rely on those for "safe" refactors.)

lewis6991 replied:
Hmm ok. The main reason I mention is that I do think `.*` is generally not needed for almost all filetypes. Usually the convention will be something like `tool/file.conf` which then gets the pattern `tool/.*.conf`, even though `tool/a/b/c.conf` probably shouldn't be filetype matched.

clason replied:
What is the actual proposed change here (in regex and pattern)?

clason replied:
(Also, note how I fucked up even _this_ pattern...)

lewis6991 replied:
Nothing concrete atm, I'm just trying to gauge what the design space is and how we can go about changing some of these patterns.

But mostly it'll be replacing `.*` with something more specific and adding more anchors (`^` and `$`). Both these things will enable breaking down patterns into multiple smaller patterns that can be structured and shared.

clason replied:
The internal representation is completely open; the only constraint is matching observable Vim behavior and being able to reasonably port Vim filetypes.

lewis6991 replied:
I'm suggesting that these patterns are tightened up so while `hypr/a.conf` would match as `hypr`, `hypr/a/b.conf` would not; so this would be an observable difference. My argument is that `hypr/a/b.conf` likely isn't valid so the observable vim behaviour is wrong (and slow).

clason replied:
That would be a change that would have to go through Vim first. Be prepared for arguments.

(No argument from me, although I would expect that at least for some patterns, this _would_ be valid, so a case-by-case decision will be needed.)

lewis6991 replied:
Ok sure, thanks for the advice. If I continue with this, then I'll take that route. I'll draft-PR what I've done so far at some point for reference and maybe make a tracking issue.

clason replied:
In general, this sounds to me like a direction we would want to go in. But realistically, there would be significant friction from upstream, not least due to the many filetype maintainers that will have their say. It might actually pay off to rewrite their filetype.vim along the same lines as filetype.lua (but in vim9script, of course) and make these changes as part of the rewrite. (Not that I see any of us volunteering for that, nor being able to nerd-snipe any of the usual upstream suspects...)

zeertzjq commented on test/old/testdir/test_filetype.vim:
Missing change to `lisp` line

clason committed:
vim-patch:9.1.0880: filetype: C3 files are not recognized

Problem:  filetype: C3 files are not recognized
Solution: detect '*.c3*' files as c3 filetype (Turiiya)

closes: vim/vim#16087

https://github.com/vim/vim/commit/c8dfcfc53ba5ed69b5d4e534fd7e8694de014e6a

Co-authored-by: Turiiya <34311583+ttytm@users.noreply.github.com>
